### PR TITLE
Bug fixes: torch::tensor(floating-point values) -> default dtype, and torch::tensor(integer values) ->at::kLong

### LINKED ISSUE #32234: Passing a float variable to torch::tensor constructor doesn't produce a tensor with default dtype
## 🐛 Bug

In Python, we have the following behavior:
```python
>>> torch.set_default_dtype(torch.double)
>>> a = 1.1
>>> t = torch.tensor([a, a])
>>> t.dtype
torch.float64
```

However, currently in C++, we have the following behavior:
```cpp
torch::set_default_dtype(torch::scalarTypeToTypeMeta(torch::kDouble));
float a = 1.1;
torch::tensor({a, a}).dtype()  // prints: float
```

We should fix the C++ API behavior (by returning a double tensor instead of a float tensor in the above case), to match the Python API.

cc @yf225

### PR DESCRIPTION
**Pull request by yf225**
Some of the `torch::tensor` behavior is updated to better match Python API. Fixes https://github.com/pytorch/pytorch/issues/32234.

This PR is BC-breaking in the following way:
- `torch::tensor({1.0f, 2.0f})`: float -> default dtype
- `torch::tensor(at::ArrayRef<int>({1, 2, 3}))`: int -> at::kLong
- `torch::tensor(std::vector<int>({1, 2, 3}))`: int -> at::kLong
- `torch::tensor(at::ArrayRef<float>({1.f, 2.f, 3.f}))`: float -> default dtype
- `torch::tensor(std::vector<float>({1.f, 2.f, 3.f}))`: float -> default dtype
- `torch::tensor(at::ArrayRef<double>({1., 2., 3.}))`: double -> default dtype
- `torch::tensor(std::vector<double>({1., 2., 3.}))`: double -> default dtype

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 4cd3f60e:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=4cd3f60ef32aca577686b6b91cf251f2770b3f63).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4292609) pytorch_macos_10_13_py3_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87888965) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87888965))
<details>
<summary>
<code>Jan 17 12:16:02 FAILED: test_api/CMakeFiles/test_api.dir/tensor.cpp.o</code>
</summary>

```
Jan 17 12:15:29 [2815/2991] Building CXX object test_api/CMakeFiles/test_api.dir/serialize.cpp.o 
Jan 17 12:15:29 ../test/cpp/api/serialize.cpp:397:25: warning: comparison of integers of different signs: 'int64_t' (aka 'long long') and 'std::__1::vector<at::Tensor, std::__1::allocator<at::Tensor> >::size_type' (aka 'unsigned long') [-Wsign-compare] 
Jan 17 12:15:29   for (int64_t i = 0; i < x_vec.size(); i++) { 
Jan 17 12:15:29                       ~ ^ ~~~~~~~~~~~~ 
Jan 17 12:15:29 1 warning generated. 
Jan 17 12:15:29 [2816/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor_cuda.cpp.o 
Jan 17 12:15:30 [2817/2991] Building CXX object test_api/CMakeFiles/test_api.dir/support.cpp.o 
Jan 17 12:15:47 [2818/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor_options_cuda.cpp.o 
Jan 17 12:15:48 [2819/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor_indexing.cpp.o 
Jan 17 12:16:02 [2820/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor.cpp.o 
Jan 17 12:16:02 FAILED: test_api/CMakeFiles/test_api.dir/tensor.cpp.o  
MacOSX10.13.sdk -mmacosx-version-min=10.9 -fPIE   -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wno-missing-braces -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT test_api/CMakeFiles/test_api.dir/tensor.cpp.o -MF test_api/CMakeFiles/test_api.dir/tensor.cpp.o.d -o test_api/CMakeFiles/test_api.dir/tensor.cpp.o -c ../test/cpp/api/tensor.cpp 
Jan 17 12:16:02 ../test/cpp/api/tensor.cpp:232:26: error: conversion from 'long' to 'detail::TensorDataContainer' is ambiguous 
Jan 17 12:16:02   tensor = torch::tensor(123l); 
Jan 17 12:16:02                          ^~~~ 
Jan 17 12:16:02 ../torch/csrc/api/include/torch/detail/TensorDataContainer.h:106:51: note: candidate constructor 
Jan 17 12:16:02 AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR) 
Jan 17 12:16:02                                                   ^ 
Jan 17 12:16:02 ../torch/csrc/api/include/torch/detail/TensorDataContainer.h:106:51: note: candidate constructor 
Jan 17 12:16:02 ../torch/csrc/api/include/torch/detail/TensorDataContainer.h:106:51: note: candidate constructor 
Jan 17 12:16:02 ../torch/csrc/api/include/torch/detail/TensorDataContainer.h:106:51: note: candidate constructor 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4292611) pytorch_macos_10_13_cuda10_0_cudnn7_py3_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87889123) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87889123))
<details>
<summary>
<code>Jan 17 12:16:18 FAILED: test_api/CMakeFiles/test_api.dir/tensor.cpp.o</code>
</summary>

```
Jan 17 12:15:37 [2816/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor_cuda.cpp.o 
Jan 17 12:15:51 [2817/2991] Building CXX object test_api/CMakeFiles/test_api.dir/serialize.cpp.o 
Jan 17 12:15:51 ../test/cpp/api/serialize.cpp:397:25: warning: comparison of integers of different signs: 'int64_t' (aka 'long long') and 'std::__1::vector<at::Tensor, std::__1::allocator<at::Tensor> >::size_type' (aka 'unsigned long') [-Wsign-compare] 
Jan 17 12:15:51   for (int64_t i = 0; i < x_vec.size(); i++) { 
Jan 17 12:15:51                       ~ ^ ~~~~~~~~~~~~ 
Jan 17 12:15:51 1 warning generated. 
Jan 17 12:15:52 [2818/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor_options_cuda.cpp.o 
Jan 17 12:16:00 [2819/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor_indexing.cpp.o 
Jan 17 12:16:13 [2820/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor_options.cpp.o 
Jan 17 12:16:18 [2821/2991] Building CXX object test_api/CMakeFiles/test_api.dir/tensor.cpp.o 
Jan 17 12:16:18 FAILED: test_api/CMakeFiles/test_api.dir/tensor.cpp.o  
macosx-version-min=10.9 -fPIE   -DCAFFE2_USE_GLOO -DCUDA_HAS_FP16=1 -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wno-missing-braces -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT test_api/CMakeFiles/test_api.dir/tensor.cpp.o -MF test_api/CMakeFiles/test_api.dir/tensor.cpp.o.d -o test_api/CMakeFiles/test_api.dir/tensor.cpp.o -c ../test/cpp/api/tensor.cpp 
Jan 17 12:16:18 ../test/cpp/api/tensor.cpp:232:26: error: conversion from 'long' to 'detail::TensorDataContainer' is ambiguous 
Jan 17 12:16:18   tensor = torch::tensor(123l); 
Jan 17 12:16:18                          ^~~~ 
Jan 17 12:16:18 ../torch/csrc/api/include/torch/detail/TensorDataContainer.h:106:51: note: candidate constructor 
Jan 17 12:16:18 AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR) 
Jan 17 12:16:18                                                   ^ 
Jan 17 12:16:18 ../torch/csrc/api/include/torch/detail/TensorDataContainer.h:106:51: note: candidate constructor 
Jan 17 12:16:18 ../torch/csrc/api/include/torch/detail/TensorDataContainer.h:106:51: note: candidate constructor 
Jan 17 12:16:18 ../torch/csrc/api/include/torch/detail/TensorDataContainer.h:106:51: note: candidate constructor 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h [hunk 1]
```diff
@@ -154,7 +148,7 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
 
   // NOTE: We need to handle `std::vector` explicitly instead of relying on an implicit conversion
   // to `at::ArrayRef`, otherwise the following error can be thrown when calling
-  // `torch::tensor(std::vector<double>({1.1, 2.2}))`:
+  // `torch::tensor(std::vector<int>({1, 2}))`:
```
Comment (zou3519) [inline torch/csrc/api/include/torch/detail/TensorDataContainer.h]: Why did this change?

**Review by zou3519 — APPROVED** [review]
lgtm

File: torch/csrc/api/include/torch/detail/TensorDataContainer.h (hunk 1 above)
Comment (yf225) [reply torch/csrc/api/include/torch/detail/TensorDataContainer.h]: I made a typo when I wrote this comment originally :/

**Pull request closed without merging**